core/shared/src/main/scala-2.13+/fs2/Chunk213And3Compat.scala

```diff
@@ -0,0 +1,162 @@
+    * Contrary to all methods that start with _"to"_ (e.g. {{toVector}}, {{toArray}}),
+    * this method does not copy data.
+    * As such, this method is mostly intended for `foreach` kind of interop.
+    */
+  def asSeq: Seq[O] =
+    new ChunkAsSeq(this)
+}
+
+private[fs2] final class ChunkAsSeq[+O](
+    private[fs2] val chunk: Chunk[O]
+) extends Seq[O]
+    with Serializable {
+  override def iterator: Iterator[O] =
+    chunk.iterator
+
+  override def apply(i: Int): O =
+    chunk.apply(i)
+
+  override def length: Int =
+    chunk.size
+
+  override def knownSize: Int =
+    chunk.size
+
+  override def isEmpty: Boolean =
+    chunk.isEmpty
+
+  override def reverseIterator: Iterator[O] =
+    chunk.reverseIterator
+
+  override def foreach[U](f: O => U): Unit =
+    chunk.foreach { o => f(o); () }
+
+  override def tapEach[U](f: O => U): Seq[O] = {
+    chunk.foreach { o => f(o); () }
+    this
+  }
+
+  override def copyToArray[O2 >: O](xs: Array[O2], start: Int, len: Int): Int = {
+    chunk.take(len).copyToArray(xs, start)
+    math.min(len, xs.length - start)
+  }
+
+  override def headOption: Option[O] =
+    chunk.head
+
+  override def head: O =
+    if (chunk.nonEmpty) chunk.apply(0)
+    else throw new NoSuchElementException("head of empty Seq")
+
+  override def lastOption: Option[O] =
+    chunk.last
+
+  override def last: O =
+    if (chunk.nonEmpty) chunk.apply(chunk.size - 1)
+    else throw new NoSuchElementException("tail of empty Seq")
+
+  override def take(n: Int): Seq[O] =
+    new ChunkAsSeq(chunk.take(n))
+
+  override def takeRight(n: Int): Seq[O] =
+    new ChunkAsSeq(chunk.takeRight(n))
+
+  override def to[C1](factory: Factory[O, C1]): C1 =
+    chunk.to(factory)
+
+  override def toArray[O2 >: O: ClassTag]: Array[O2] =
+    chunk.toArray
+
+  override def toList: List[O] =
+    chunk.toList
+
+  override def toVector: Vector[O] =
+    chunk.toVector
+
+  override def toString: String =
+    chunk.iterator.mkString("ChunkAsSeq(", ", ", ")")
+
+  override def zipWithIndex: Seq[(O, Int)] =
+    new ChunkAsSeq(chunk.zipWithIndex)
+
+  override def equals(that: Any): Boolean =
+    that match {
+      case thatChunkWrapper: ChunkAsSeq[_] =>
+        chunk == thatChunkWrapper.chunk
+
+      case _ =>
+        false
+    }
+
+  override val iterableFactory: SeqFactory[Seq] =
+    ArraySeq.untagged
+
+  override val empty: Seq[O] =
+    new ChunkAsSeq(Chunk.empty)
+}
+
+private[fs2] trait ChunkCompanion213And3Compat { self: Chunk.type =>
+
+  protected def platformIterable[O](i: Iterable[O]): Option[Chunk[O]] =
+    i match {
+      case a: ArraySeq[O]   => Some(arraySeq(a))
+      case w: ChunkAsSeq[O] => Some(w.chunk)
+      case _                => None
+    }
+
+  /** Creates a chunk backed by an immutable `ArraySeq`. */
+  def arraySeq[O](arraySeq: ArraySeq[O]): Chunk[O] = {
+    val arr = arraySeq.unsafeArray.asInstanceOf[Array[O]]
+    array(arr)(ClassTag[O](arr.getClass.getComponentType))
+  }
```

Comment:
I see that you copied it from the previous implementation to consolidate it in one file. However, perhaps such an implementation does not make a lot of sense anymore since there's `indexedSeq` method already and `ArraySeq` inherits to `IndexedSeq`. So I guess you could simply redirect this method to `indexedSeq` instead.

Reply:
Hmm, I think it still makes sense. This is getting at the underlying `Array` which is much more optimized to work with e.g. see `ArraySlice`.

Reply:
Maybe.. But in that case I feel it would make sense to add such an optimization to the `indexedSeq` itself, i.e. if a seq is an `ArraySeq`, then make an `ArraySlice` out of it.

Then redirect `arraySeq` to `indexedSeq` :) 

But it seems to be out of scope for this PR anyway.

Reply:
@satorg `indexedSeq` is AFAIK generic to all **Scala** versions, so it can check for `arraySeq`

Reply:
Take a look at the `platformIterable` method, which is used in the implementation of `iterable`, which in turn `seq` delegates to. We could unwind this a bit, deprecating the public facing methods that simply delegate to the type checking uber method.

Reply:
@mpilquist so you are suggesting that we deprecate the `arraySeq` factories and rather let `iterable` call the optimized impl?
I don't have an issue with that, should we also deprecate `indexedSeq` and `seq` as well and rather just have a single `iterable` factory?

Reply:
We could, yeah. The various specific constructors, like `arraySeq` and `indexedSeq` and `vector` all wrap direct construction of a corresponding `Chunk` subtype. Whereas `iterable` does type casing to pick the most efficient subtype. Then `seq` breaks the first rule and delegates to `iterable`.

Moving forward, one option is deprecating all of those and creating `Chunk.collection` (or `fromCollection` or maybe even just `from`) which uses the current implementation of `iterable`.